### PR TITLE
Added proper baritone support, and fixed villager interactions on servers with anticheat

### PR DESCRIPTION
**Pull request by Mark7888**
- Added a proper baritone support to the current version based on #29 .
- Added an option to timeout waiting for interraction packets
- Added an option to look at the villager when interracting (On 2b2t the interraction is getting cancelled by the anticheat if the player is not looking at the villager directly)

### CODE REVIEW

**Comment by maxsupermanhd** [comment]
Hi, thank you for your time contributing. Unfortunately, you probably didn't look into changes requested to #29 and I do not blame you for that and appreciate the effort. You clearly went out of the way to do the *proper* thing and I like where it is going, moreover there seem to be real need for this.

Let me rephrase what I said in #29: what you acquire from `getCommandManager()` is intended to be used by UI mods in order to direct user input to the baritone. Probably `GoalPlace` from `baritone.process` should be used (through `ICustomGoalProcess` maybe even) just like many modules in meteor itself do, it is *correct* way of implementing it. Commands are for *users*. Processes are for mods.

Great example of baritone usage that should be implemented can be find in InfinityMiner module here: https://github.com/MeteorDevelopment/meteor-client/blob/master/src/main/java/meteordevelopment/meteorclient/systems/modules/world/InfinityMiner.java

*Side note: if we are looking onto the villager to bypass anticheats we perhaps should not "jump look" and mimic mouse movement via multiple consecutive look at packets*

*Also please correct spelling of "interract".*

**Review by maxsupermanhd — CHANGES_REQUESTED** [review]
Still using commands...

**Comment by Mark7888** [comment]
Hey. Yep, I totally missed that part. Will rewrite that way, and push it here.

**Comment by Mark7888** [comment]
Hi. 
Yesterday I have managed to make the baritone place properly work with [FillSchematic](https://baritone.leijurv.com/baritone/api/schematic/CompositeSchematicEntry.html).

Meanwhile I might have found a proper way of handling the mouse movement when looking at the villager. Will try to add it in a couple of hours.

File: src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java [hunk 1]
```diff
@@ -780,19 +853,66 @@ private void onTick(TickEvent.Pre event) {
...
+                    int x = rollingBlockPos.getX();
+                    int y = rollingBlockPos.getY();
+                    int z = rollingBlockPos.getZ();
+
+                    // Use Baritone to place the lectern block
+                    IBaritone baritone = BaritoneAPI.getProvider().getPrimaryBaritone();
+                    BetterBlockPos lecternPosition = new BetterBlockPos(x, y, z);
+                    BlockOptionalMeta type = new BlockOptionalMeta(Blocks.LECTERN);
+                    ISchematic schematic = new FillSchematic(1, 1, 1, type);
+                    baritone.getBuilderProcess().build("Fill", schematic, lecternPosition);
```
Comment (maxsupermanhd) [inline src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:882]: In docs for `build()` it says "Requests" and other API examples I seen ensure baritone is stopped before scheduling anything else, not sure if it is needed here but still.

File: src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java [hunk 2]
```diff
@@ -821,10 +941,19 @@ private void onTick(TickEvent.Pre event) {
                     return;
                 }
                 if (rollingVillager.getVillagerData().getProfession() != VillagerProfession.NONE) {
-                    currentState = State.ROLLING_WAITING_FOR_VILLAGER_TRADES;
+                    currentState = State.ROLLING_WAITING_FOR_VILLAGER_TRADES;
```
Comment (maxsupermanhd) [inline src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:944]: ???

**Review by maxsupermanhd — APPROVED** [review]

File: src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java (hunk 1 above)
Comment (Mark7888) [reply src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:882]: What API docs are you using? I can't seem to find it. I used [this](https://baritone.leijurv.com/baritone/api/process/IBuilderProcess.html) API docs, and it gives no detailed description about this function.

I can do a [cancel](https://github.com/cabaletta/baritone/blob/849072f8ca99e5d028ec587e64952c2fdd878653/src/main/java/baritone/command/defaults/ForceCancelCommand.java#L40) before scheduling the Lectern placement, but I got this fill solution from the original baritone repository's `select fill` command [here](https://github.com/cabaletta/baritone/blob/849072f8ca99e5d028ec587e64952c2fdd878653/src/main/java/baritone/command/defaults/SelCommand.java#L124), and I dont think they cancel anything before scheduling the `build()`

File: src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java (hunk 2 above)
Comment (Mark7888) [reply src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:944]: I have no idea how that got there, probably started writing something, and than deleted it, and that stayed there. Will delete it in the next commit.

File: src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java (hunk 1 above)
Comment (maxsupermanhd) [reply src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:882]: image

I guess then the only way to ensure is to test it out

Comment (Mark7888) [reply src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:882]: I will add cancel before placing the block, because added a functionality to use baritone [ILookBehavior](https://baritone.leijurv.com/baritone/api/behavior/ILookBehavior.html) to look at the villager (will also add a non-bariton variant), if baritone is enabled, and sometimes even that look request can make baritone not place the block.

Comment (maxsupermanhd) [reply src/main/java/maxsuperman/addons/roller/modules/VillagerRoller.java:882]: Oh, right, ILookBehavior is something that could be used for smooth looking towards villager, good job

**Comment by Mark7888** [comment]
Added some random noise to the `yaw` and `pitch` coordinates when looking at the villager. Also fi baritone is enabled, we can use that to change target.

Tested with 3 different anticheats, and non of them triggered on the "fast mouse movement" when changing rotation. But one of them triggered on the original block breaking method, so I will also add a baritone way of breaking the lectern.

**Comment by maxsupermanhd** [comment]
Very cool, much appreciate, I would like to do some further testing on edge cases and then merge it with minor version bump. Might take couple of days. Probably good idea to enable dev build and merge it so that others can test it with ease too.

**Comment by Mark7888** [comment]
- Moved the block break/placement logic to function.
- Added the block break with baritone.
- Renamed the option from `baritone-place` to `use-baritone` (and updated the description)

**Comment by maxsupermanhd** [comment]
So I did a bit of testing and so far (except usual baritone junk) only problem is that you crash on addon initialize because you use baritone directly.

In meteor modules that are using baritone are either initialized if baritone is present or not initialized at all to avoid no class def error.

Unfortunately I am unsure on how to properly solve it, perhaps some "BaritoneHelper" class that would only be initialized and non-nulled on init if baritone is present and then a bunch of conditional use of this helper if it is null or not.

I think I will hopefully figure something out soon enough.

**Comment by Big-Iron-Cheems** [comment]
In meteor we have an [helper class](https://github.com/MeteorDevelopment/meteor-client/blob/master/src/main/java/meteordevelopment/meteorclient/pathing/BaritoneUtils.java) you can query at runtime to see if Baritone is present.

Example usage in the [Modules](https://github.com/MeteorDevelopment/meteor-client/blob/3f8efd8c08a309ea7279c1c9116d0aad4e7b8870/src/main/java/meteordevelopment/meteorclient/systems/modules/Modules.java#L569) class.

**Comment by maxsupermanhd** [comment]
> you can query at runtime to see if Baritone is present

Right but how do I instantiate a module that references baritone classes if baritone is not present tho?

**Comment by Big-Iron-Cheems** [comment]
Well, hiding settings based on Baritone's presence is possible.
What remains later is adding a bunch of checks in any place that would call something related to baritone.

It would look a bit messy, but I think it could work. Got a class I can look at to try and think about it?

**Comment by maxsupermanhd** [comment]
> Got a class I can look at to try and think about it?

Well, in particular, if you build this gentleman’s branch and try to launch it without having baritone loaded it will crash with no class def error on init, even though it just mentions the classes inside private method of a module class. (basically will be confused to see BetterBlockPos here for example https://github.com/maxsupermanhd/meteor-villager-roller/pull/44/files#diff-850f07a494dca95a073ce5fa82e575fcb623a0ad33e752b5a58669407158b9a1R833 even though this function is never to be called in runtime)

**Comment by Big-Iron-Cheems** [comment]
Ok I understand the issue now, it's really simple.
In your gradle dependencies Baritone is added as `modCompileOnly`, which works fine in a dev env, but expects the actual mod JAR in normal gameplay.
Whenever the `VillagerRoller` class is trying to be instantiated all imports must be validated, which is why this crashes without the Baritone JAR. It's not a matter of code that has been ran, you're not at that step yet.

Now, the solution could be to create some sort of `BlockPlacer` interface, and implement two versions: 
- a `DefaultBlockPlacer` that uses the current methods of block placing/breaking/looking etc
- and a `BaritoneBlockPlacer` that uses the corresponding Baritone logic

This would then be managed by some `BlockPlacerFactory` that at runtime uses our Baritone check and returns the corresponding instance in the `VillagerRoller` class, in which we can now do something like:
```java
private final BaritoneBlockPlacer blockPlacer = BlockPlacerFactory.getBlockPlacer();
blockPlacer.placeBlock(...); // or whatever methods you define in the interface
```

TL;DR move the Baritone imports out of your VillagerRoller class to avoid runtime crashes and properly abstract the actions based on the `BlockPlacer` type you determined.

**Comment by maxsupermanhd** [comment]
Yeah I imagined something like this will be needed, very annoying...

Thank you for proper explanation. I guess time to make some kind of abstract BaritonableAction, wish something like that existed in meteor already

**Comment by Mark7888** [comment]
Well this is strange, because I have tested this without baritone, even added checks to ensure none of the baritone code runs when baritone is not installed, and worked for a while.
Sadly I did not test this case after moving baritone code to a separate function, that might be the thing breaking it, because before that, when I only had the baritone-break, it worked, so if baritone is not installed, it automatically disabled the functionality.

I'm very sorry for the inconvenience, I didn't really know how this part of modding worked. I know Java because I've written a lot of spigot plugins before, but not mods.

I will read this thread through and try to figure it out, if it wont be solved by anyone else by then.

**Comment by Big-Iron-Cheems** [comment]
I already sent Max a patch file with all the proposed changes and it seems to have fixed it.
When he has time he'll probably commit it.
Here it is in case you want to look through my work: [Baritone_integration.patch](https://github.com/user-attachments/files/18815343/Baritone_integration.patch)

**Comment by Mark7888** [comment]
Oh, okay. Great!
Will take a look at it later, thank you!

**Comment by maxsupermanhd** [comment]
not sure how to properly do it so I guess will just slap it onto master and see what happens
